### PR TITLE
[ALLUXIO-2913]Separate LineageMaster.java into an interface class and an implementation class

### PR DESCRIPTION
**Pull request by cmuyaol**
https://alluxio.atlassian.net/browse/ALLUXIO-2913

### CODE REVIEW

**Comment by gpang** [comment]
Jenkins, test this please

File: core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+   * @param job the job
+   * @return the id of the created lineage
+   * @throws InvalidPathException if the path to the input file is invalid
+   * @throws FileAlreadyExistsException if the output file already exists
+   * @throws BlockInfoException if fails to create the output file
+   * @throws AccessControlException if the permission check fails
+   * @throws FileDoesNotExistException if any of the input files do not exist
+   */
+  @Override
+  public synchronized long createLineage(List<AlluxioURI> inputFiles, List<AlluxioURI> outputFiles,
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java]: for `Override` methods, we don't need javadoc

File: core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+   * Deletes a lineage.
+   *
+   * @param lineageId id the of lineage
+   * @param cascade the flag if to delete all the downstream lineages
+   * @return true if the lineage is deleted, false otherwise
+   * @throws LineageDoesNotExistException the lineage does not exist
+   * @throws LineageDeletionException the lineage deletion fails
+   */
+  @Override
+  public synchronized boolean deleteLineage(long lineageId, boolean cascade)
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java]: ditto

File: core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+   * @param ttl the TTL
+   * @param ttlAction action to perform on ttl expiry
+   * @return the id of the reinitialized file when the file is lost or not completed, -1 otherwise
+   * @throws InvalidPathException the file path is invalid
+   * @throws LineageDoesNotExistException when the file does not exist
+   * @throws AccessControlException if permission checking fails
+   * @throws FileDoesNotExistException if the path does not exist
+   */
+  @Override
+  public synchronized long reinitializeFile(String path, long blockSizeBytes, long ttl,
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java]: ditto

File: core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+  /**
+   * Reports a file as lost.
+   *
+   * @param path the path to the file
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws AccessControlException if permission checking fails
+   * @throws InvalidPathException if the path is invalid
+   */
+  @Override
+  public synchronized void reportLostFile(String path)
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java]: ditto

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
@cmuyaol 

File headers are missing for the following files
```
[WARNING] Missing header in: /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/master/src/test/java/alluxio/master/lineage/LineageMasterTest.java
[WARNING] Missing header in: /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/master/src/main/java/alluxio/master/lineage/DefaultLineageMaster.java
[WARNING] Missing header in: /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/master/src/main/java/alluxio/master/lineage/LineageMaster.java
```
You can copy the header from other files and add them there

**Comment by cmuyaol** [comment]
Hi, is there still any problem in this PR? @apc999

File: core/server/master/src/main/java/alluxio/master/lineage/LineageMasterClientRestServiceHandler.java [hunk 5]
```diff
@@ -68,7 +68,7 @@ public LineageMasterClientRestServiceHandler(@Context ServletContext context) {
     // Poor man's dependency injection through the Jersey application scope.
     mLineageMaster = ((MasterProcess) context
         .getAttribute(MasterWebServer.ALLUXIO_MASTER_SERVLET_RESOURCE_KEY))
-        .getMaster(LineageMaster.class);
+        .getMaster(DefaultLineageMaster.class);
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/LineageMasterClientRestServiceHandler.java]: revert this change

File: core/server/master/src/main/java/alluxio/master/lineage/LineageMasterFactory.java [hunk 6]
```diff
@@ -50,10 +50,10 @@ public String getName() {
   @Override
   public LineageMaster create(MasterRegistry registry, JournalFactory journalFactory) {
     Preconditions.checkArgument(journalFactory != null, "journal factory may not be null");
-    LOG.info("Creating {} ", LineageMaster.class.getName());
+    LOG.info("Creating {} ", DefaultLineageMaster.class.getName());
     FileSystemMaster fileSystemMaster = registry.get(FileSystemMaster.class);
-    LineageMaster lineageMaster = new LineageMaster(fileSystemMaster, journalFactory);
-    registry.add(LineageMaster.class, lineageMaster);
+    LineageMaster lineageMaster = new DefaultLineageMaster(fileSystemMaster, journalFactory);
+    registry.add(DefaultLineageMaster.class, lineageMaster);
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/LineageMasterFactory.java]: `registry.add(LineageMaster.class, lineageMaster);`
Here we register implementations based on its type (interface)

File: core/server/master/src/main/java/alluxio/master/lineage/LineageMasterFactory.java [hunk 7]
```diff
@@ -50,10 +50,10 @@ public String getName() {
   @Override
   public LineageMaster create(MasterRegistry registry, JournalFactory journalFactory) {
     Preconditions.checkArgument(journalFactory != null, "journal factory may not be null");
-    LOG.info("Creating {} ", LineageMaster.class.getName());
+    LOG.info("Creating {} ", DefaultLineageMaster.class.getName());
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/LineageMasterFactory.java]: revert this change

File: core/server/master/src/main/java/alluxio/master/lineage/LineageMasterClientServiceHandler.java [hunk 8]
```diff
@@ -61,7 +61,7 @@
   /**
    * Creates a new instance of {@link LineageMasterClientServiceHandler}.
    *
-   * @param lineageMaster the {@link LineageMaster} the handler uses internally
+   * @param lineageMaster the {@link DefaultLineageMaster} the handler uses internally
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/LineageMasterClientServiceHandler.java]: can you revert all the change from `LineageMaster` to `DefaultLineageMaster` in Javadoc like this, and below?

File: core/server/master/src/test/java/alluxio/master/lineage/LineageMasterTest.java [hunk 9]
```diff
@@ -73,9 +73,9 @@ public void before() throws Exception {
     mRegistry.add(FileSystemMaster.class, mFileSystemMaster);
     ThreadFactory threadPool = ThreadFactoryUtils.build("LineageMasterTest-%d", true);
     mExecutorService = Executors.newFixedThreadPool(2, threadPool);
-    mLineageMaster = new LineageMaster(mFileSystemMaster, factory,
+    mLineageMaster = new DefaultLineageMaster(mFileSystemMaster, factory,
         ExecutorServiceFactories.constantExecutorServiceFactory(mExecutorService));
-    mRegistry.add(LineageMaster.class, mLineageMaster);
+    mRegistry.add(DefaultLineageMaster.class, mLineageMaster);
```
Comment (apc999) [inline core/server/master/src/test/java/alluxio/master/lineage/LineageMasterTest.java]: `mRegistry.add(LineageMaster.class, mLineageMaster);

File: tests/src/test/java/alluxio/master/lineage/LineageMasterClientRestApiTest.java [hunk 10]
```diff
@@ -91,7 +91,7 @@ public void createLineage() throws Exception {
   @Test
   @Config(confParams = {PropertyKey.Name.USER_LINEAGE_ENABLED, "true"})
   public void deleteLineage() throws Exception {
-    LineageMaster lineageMaster = mMasterProcess.getMaster(LineageMaster.class);
+    LineageMaster lineageMaster = mMasterProcess.getMaster(DefaultLineageMaster.class);
```
Comment (apc999) [inline tests/src/test/java/alluxio/master/lineage/LineageMasterClientRestApiTest.java]: `getMaster(LineageMaster.class);`

File: tests/src/test/java/alluxio/master/lineage/LineageMasterClientRestApiTest.java [hunk 11]
```diff
@@ -108,7 +108,7 @@ public void deleteLineage() throws Exception {
   public void getLineageInfoList() throws Exception {
     AlluxioURI input = new AlluxioURI("/input");
     AlluxioURI output = new AlluxioURI("/output");
-    LineageMaster lineageMaster = mMasterProcess.getMaster(LineageMaster.class);
+    LineageMaster lineageMaster = mMasterProcess.getMaster(DefaultLineageMaster.class);
```
Comment (apc999) [inline tests/src/test/java/alluxio/master/lineage/LineageMasterClientRestApiTest.java]: `getMaster(LineageMaster.class);`

File: core/server/master/src/test/java/alluxio/master/lineage/LineageMasterTest.java [hunk 12]
```diff
@@ -48,7 +48,7 @@
 import java.util.concurrent.ThreadFactory;
 
 /**
- * Unit tests for {@link LineageMaster}.
+ * Unit tests for {@link DefaultLineageMaster}.
```
Comment (apc999) [inline core/server/master/src/test/java/alluxio/master/lineage/LineageMasterTest.java]: can you revert all the change from `LineageMaster` to `DefaultLineageMaster` in Javadoc like this in this file?

File: core/server/master/src/main/java/alluxio/master/lineage/LineageMasterFactory.java [hunk 13]
```diff
@@ -26,7 +26,7 @@
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * Factory to create a {@link LineageMaster} instance.
+ * Factory to create a {@link DefaultLineageMaster} instance.
```
Comment (apc999) [inline core/server/master/src/main/java/alluxio/master/lineage/LineageMasterFactory.java]: `{@link LineageMaster}` -> `{@link DefaultLineageMaster}`
